### ISSUE TITLE
SA1642 (constructor summary text) special case for incorrect class name

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#415]
When SA1642 is reported, it is always reported for the entire `<summary>` element. In each of the following scenarios, a more specific location should be used for reporting the warning.

1. When the name of the class is not wrapped in a `<see>` element, but the text is otherwise correct. For example, the warning reported for the following code should only underline the word `Customer`:

    ```csharp
    /// <summary>
    /// Initializes a new instance of the Customer class.
    /// </summary>
    public Customer() { }
    ```

1. As in the first case, but now considering an incorrect class name. For example:

    ```csharp
    /// <summary>
    /// Initializes a new instance of the NotTheRightType class.
    /// </summary>
    public Customer() { }
    ```

1. When the name of the class specified in the `cref` attribute does not match the name of the current class, but the text is otherwise correct. In this case, the warning should be reported for the `cref` value instead of the entire `<summary>` element. For example:

    ```csharp
    /// <summary>
    /// Initializes a new instance of the <see cref="NotTheRightType"/> class.
    /// </summary>
    public Customer() { }
    ```

**Comment by sharwell** [comment]
I initially assigned @pdelvo since he did a great job with the initial implementation. I'll take a look soon if you don't have time.

**Comment by sharwell** [comment]
:bulb: The code fix for each of the above cases should replace the current type reference with the corrected type reference, as opposed to inserting the entire standard prefix.

**Comment by sharwell** [comment]
:question: I also noticed a case like the following:

> Initializes a new instance of the &lt;see cref="MyType"/&gt; with A and B.

Notice that everything is correct except the word "class" has been omitted. Should we special case handling for code that is *close* but not exactly correct? If so, how close is close?

**Comment by pdelvo** [comment]
This is getting complicated. Calculating the [Levenshtein distance](http://en.wikipedia.org/wiki/Levenshtein_distance) and having an upper limit would make this rule almost useless. If I have a spelling mistake in there the distance would be very small and I would not get a diagnostic. But we could add a few special cases like the one you mentioned.

**Comment by sharwell** [comment]
I tend to agree but wanted to ask the question publicly before just deciding :+1:

**Comment by sharwell** [comment]
Doh. I thought I made a new issue for the "closeness" question.

**Comment by sharwell** [comment]
The incorrect class name situation is common and relatively simple to detect/correct.

**Comment by pdelvo** [comment]
I am working on this. It is getting a bit complicated but my solution is able to also detect spelling mistakes and changing the comment instead of adding a new one

**Comment by sharwell** [comment]
Well that would be cool :+1: It's OK if it doesn't work out too.
